### DISCUSSION TITLE
How do you create a view with group by?

### DISCUSSION THREAD

**Discussion by Foxsterc** [simolus3/drift#2129] [General]
Creating a view like

```dart
abstract class OrderView extends View {
  
  Orders get orders;

  
  Expression<int> get total => orders.customerId.count();

  @override
  Query as() => select([orders.customerId]).from(orders).addColumns([total])
    ..groupBy([orders.customerId]);
  
}
```

generates invalid code

```
  @override
  Query? get query =>
      (attachedDatabase.selectOnly(orders)..addColumns($columns))
          .addColumns([total])
        ..groupBy([orders.customerId]);
```

Is it possible to use grouping in a View?

**Comment by simolus3** [comment]
You don't need to call `addColumns` in a view, just do this:

```dart
abstract class OrderView extends View {
  
  Orders get orders;

  
  Expression<int> get total => orders.customerId.count();

  @override
  Query as() => select([orders.customerId, total]).from(orders)
    ..groupBy([orders.customerId]);
  
}
```

**Reply by Foxsterc** [reply]
Thanks for the reply. I did try that but it gives...

<img width="736" alt="image" src="https://user-images.githubusercontent.com/13538417/200162086-1c2a49d3-0dc2-47d7-8613-b72917c038c7.png">

**Reply by simolus3** [reply]
Hm yeah, this is not ideal :/ I think the workaround should work if you add an empty `join` instead, like this:

```dart
abstract class OrderView extends View {
  
  Orders get orders;

  
  Expression<int> get total => orders.customerId.count();

  @override
  Query as() => select([orders.customerId]).from(orders).join([])
    ..groupBy([orders.customerId]);
  
}
```

**Reply by Foxsterc** [reply]
Yes, this works fine! Thanks for the help.

**Comment by appsowner** [comment]
@simolus3 how create the next select group by with drift ?......Thanks

 SELECT 
 t0.fechaVenta,
 t0.idventa,
 user_id,
 "t3".name username,
 "t2".dni,
 "t2".nombre nombrecliente,
 SUM(precio) totalPrecio,
 SUM(descuento) totalDescuento,
 SUM(precio-descuento) totalVenta,
 count(articulo_id) totalProductos
 FROM "ventas" "t0" INNER JOIN "clientes" "t2" ON "t2"."id" = "t0"."client_id" INNER JOIN "usuarios" "t3" ON "t3"."id" = "t0"."user_id"
 group by   
 t0.fechaVenta,
 t0.idventa,
 user_id,
 "t3".name,
 "t2".dni,
 "t2".nombre
 order by 1

**Comment by appsowner** [comment]
i made it!!

code example!

```dart
abstract class VentasAcumView extends View {
  Ventas get ventas;
  Articulos get articulos;
  Clientes get clientes;
  Usuarios get usuarios;

  Expression<int> get total => ventas.precio.sum();
  Expression<int> get totalprod => ventas.idventa.count();
  @override
  Query as() => select([
        ventas.fechaVenta,
        ventas.idventa,
        ventas.userId,
        usuarios.name,
        clientes.dni,
        clientes.nombre,
        totalprod,
        total
      ]).from(ventas).join([
        innerJoin(clientes, clientes.id.equalsExp(ventas.clientId)),
        innerJoin(usuarios, usuarios.id.equalsExp(ventas.userId))
      ])
        ..groupBy([
          ventas.fechaVenta,
          ventas.idventa,
          ventas.userId,
          ventas.userId,
          usuarios.name,
          clientes.dni,
          clientes.nombre
        ]);
}
```

but in process generate file xxx.g.dart, the next issue

in more two inner join, no generate  comma between sentence...example 

```
 (attachedDatabase.selectOnly(ventas)..addColumns($columns)).join([
        innerJoin(clientes, clientes.id.equalsExp(ventas.clientId)).         => fault comma
        innerJoin(usuarios, usuarios.id.equalsExp(ventas.userId))
```

other issue is not generated comma between fields in group by..
other issue is not generated in sentenced group by .. for example......
```
 ])
        ..groupBy([
```

regards!!
